### ISSUE TITLE
finetuning support for pytorch version

### ISSUE THREAD

**Issue by huyz1117** [google-research/timesfm#153]
Will you release the finetuning code for pytorch version?

**Comment by siriuz42** [comment]
Not in our roadmap at this moment. That said you'd need a data loader and a training loop and that should be it. The data loader needs to be customized to work on your own datasets, and that's one of the reasons we don't have off-the-shelf ft code.

**Comment by Jon-tao** [comment]
@siriuz42 thank for your return. I want to check that the finetuning can not be run in pytorch?

**Comment by doitslow** [comment]
@siriuz42 
Just a quick check, should I set return_forecast_on_context=True for finetuning? I presume it has been set to True for pretraining, but would it be necessary for finetuning?

**Comment by ani0135** [comment]
@doitslow if you have figured out how to do finetuning on pytorch, can you please share your code?

**Comment by doitslow** [comment]
> @doitslow if you have figured out how to do finetuning on pytorch, can you please share your code?

I can't share code. But here is what I did: 
 - DATA: just chop your series into pieces of seen_points + pred_points, convert input data into format that pytorch_patched_decoder.decode accepts following what timesfm_base._preprocess does; 
 - MODEL: instantiate your model using pytorch_patched_decoder.PatchedTimeSeriesDecoder
and, fill up other procedure codes to get your experiment running.

**Comment by ani0135** [comment]
Thank you for your help. Can you please explain what are they trying to do here in below code? [timesfm_link](https://github.com/google-research/timesfm/blob/ff2a350ba317f462f62485b523d16198a2508c22/src/timesfm/timesfm_torch.py#L122C5-L135C1)
Also, do I need to do what they did after getting mean_output, full_output because it has no gradient invloved?

**Comment by doitslow** [comment]
> @doitslow Thank you for your help. Can you please explain what are they trying to do here in below code? [timesfm_link](https://github.com/google-research/timesfm/blob/ff2a350ba317f462f62485b523d16198a2508c22/src/timesfm/timesfm_torch.py#L122C5-L135C1) Also, do I need to do what they did after getting mean_output, full_output because it has no gradient invloved?

I believe these lines of code are for selecting batched data. The decoder need 3 inputs: actual sequence, padding, frequency.

**Comment by doitslow** [comment]
@ani0135 You would need mean_output for loss calculation.

**Comment by ani0135** [comment]
Thank you.
________________________________

**Comment by mayuanyang** [comment]
anyone got some pytorch fine tuning codes to share?
